### ISSUE TITLE
Inconsistent undefined-variable reporting

### ISSUE THREAD

**Issue by jolaf** [PyCQA/pylint#5399]
### Bug description

The following code:
```python3
# pylint: disable=multiple-statements, no-self-use, unused-argument, used-before-assignment
class A:
    def x(self, o: A) -> bool: ...
    def z(self) -> A: ...
    def t(self) -> A: ...
```
produces only 1 error report instead of 3.

### Command used

```shell
pylint test.pyi
```

### Pylint output

```shell
test.pyi:5: [E0602(undefined-variable), A.z] Undefined variable 'A'
```

### Expected behavior

Error should be reported in lines 3, 4 and 5.

### Pylint version

```shell
pylint 2.12.1
astroid 2.9.0
Python 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
```

### OS / Environment

Ubuntu 20.04.3 LTS \n \l

Linux ubuntu 5.4.0-90-generic #101-Ubuntu SMP Fri Oct 15 20:00:55 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by jacobtylerwalls** [comment]
Might be the same as #5326, and in fact, I would be willing to rephrase it if you can help me :-).

**Comment by jacobtylerwalls** [comment]
I think what they share is the second instance of the name "consuming" the first instance where the first instance is a type annotation.

**Comment by jolaf** [comment]
> Might be the same as #5326, and in fact, I would be willing to rephrase it if you can help me :-).
How can I help?

**Comment by cdce8p** [comment]
Pylint doesn't work too well for `pyi` files at the moment 😕
I would recommend a type checker for these cases. E.g. `mypy` or `pyright`.
